vpnglean/glean_parser_ext/templates/cpp.jinja2

```diff
@@ -36,33 +34,45 @@ namespace mozilla::glean {
 {%- macro extra_keys_with_types(obj, name, suffix) -%}
 struct {{ obj.name|Camelize }}{{ suffix }} {
   {% for item, type in obj|attr(name) %}
-  mozilla::Maybe<{{type|extra_type_name}}> {{ item|camelize }};
+  QVariant _{{ item|camelize }} = QVariant();
   {% endfor %}
 
-  Tuple<nsTArray<nsCString>, nsTArray<nsCString>> ToFfiExtra() const {
-    nsTArray<nsCString> extraKeys;
-    nsTArray<nsCString> extraValues;
+  FfiExtra ToFfiExtra() const {
+    QVector<const char*> extraKeys;
+    QVector<const char*> extraValues;
     {% for item, type in obj|attr(name) %}
-    if ({{item|camelize}}) {
-      extraKeys.AppendElement()->AssignASCII("{{item}}");
+    if (_{{item|camelize}}.canConvert<{{type|extra_type_name}}>()) {
```

Comment:
These underscores address the same issue as the `r#` on Rust. In C++ we don't have the variable name escaping though, so this is worse because it requires folks to actually add the underscore when recording. 

This issue showed up for a `errno` extra key we have on the VPN metrics. I am happy to file a ticket on the Glean side to address this in a possible nicer way? Let me know.

Reply:
Weird, it shouldn't care about identifiers so long as they're not keywords, since they're namespaced (this is C++ after all). What's the error you get?

Reply:
This is the error message:

```
/home/brizental/Projects/mozilla-vpn-client/build/vpnglean/metrics.h:352:32: error: invalid pure specifier (only ‘= 0’ is allowed) before ‘(’ token
  352 |       QVariant errno = QVariant();
      |                                ^
In file included from /usr/include/c++/11/cerrno:42,
                 from /usr/include/c++/11/ext/string_conversions.h:44,
                 from /usr/include/c++/11/bits/basic_string.h:6608,
                 from /usr/include/c++/11/string:55,
                 from /usr/include/c++/11/bits/locale_classes.h:40,
                 from /usr/include/c++/11/bits/ios_base.h:41,
                 from /usr/include/c++/11/streambuf:41,
                 from /usr/include/c++/11/bits/streambuf_iterator.h:35,
                 from /usr/include/c++/11/iterator:66,
                 from /home/brizental/Qt/6.2.4/gcc_64/include/QtCore/qcontainertools_impl.h:54,
                 from /home/brizental/Qt/6.2.4/gcc_64/include/QtCore/qarraydataops.h:45,
                 from /home/brizental/Qt/6.2.4/gcc_64/include/QtCore/qarraydatapointer.h:43,
                 from /home/brizental/Qt/6.2.4/gcc_64/include/QtCore/qbytearray.h:47,
                 from /home/brizental/Qt/6.2.4/gcc_64/include/QtCore/qstringview.h:58,
                 from /home/brizental/Qt/6.2.4/gcc_64/include/QtCore/qchar.h:666,
                 from /home/brizental/Qt/6.2.4/gcc_64/include/QtCore/qstring.h:50,
                 from /home/brizental/Qt/6.2.4/gcc_64/include/QtQml/qjsvalue.h:44,
                 from /home/brizental/Qt/6.2.4/gcc_64/include/QtQml/QJSValue:1,
                 from /home/brizental/Projects/mozilla-vpn-client/src/addons/addon.h:8,
                 from /home/brizental/Projects/mozilla-vpn-client/src/addons/manager/addonmanager.h:8,
                 from /home/brizental/Projects/mozilla-vpn-client/src/mozillavpn.cpp:5:
/home/brizental/Projects/mozilla-vpn-client/build/vpnglean/metrics.h: In member function ‘mozilla::glean::FfiExtra mozilla::glean::sample::AuthenticationInappErrorExtra::ToFfiExtra() const’:
/home/brizental/Projects/mozilla-vpn-client/build/vpnglean/metrics.h:361:13: error: passing ‘const mozilla::glean::sample::AuthenticationInappErrorExtra’ as ‘this’ argument discards qualifiers [-fpermissive]
  361 |         if (errno.canConvert<QString>()) {
      |             ^~~~~
/home/brizental/Projects/mozilla-vpn-client/build/vpnglean/metrics.h:352:16: note:   in call to ‘QVariant* mozilla::glean::sample::AuthenticationInappErrorExtra::__errno_location()’
  352 |       QVariant errno = QVariant();
      |                ^~~~~
/home/brizental/Projects/mozilla-vpn-client/build/vpnglean/metrics.h:363:30: error: passing ‘const mozilla::glean::sample::AuthenticationInappErrorExtra’ as ‘this’ argument discards qualifiers [-fpermissive]
  363 |           extraValues.append(errno.toString().toUtf8().constData());
      |                              ^~~~~
/home/brizental/Projects/mozilla-vpn-client/build/vpnglean/metrics.h:352:16: note:   in call to ‘QVariant* mozilla::glean::sample::AuthenticationInappErrorExtra::__errno_location()’
  352 |       QVariant errno = QVariant();
      |                ^~~~~
gmake[2]: *** [src/CMakeFiles/mozillavpn.dir/build.make:1847: src/CMakeFiles/mozillavpn.dir/mozillavpn.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1102: src/CMakeFiles/mozillavpn.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

Reply:
Well that's weird. Lemme dig.

...oh dear.

Okay, TIL, `errno` is actually [a preprocessor macro in C](https://en.cppreference.com/w/c/error/errno), meaning that it doesn't know or care about scoping. It's essentially a keyword like `int` or `function`. 

...is it wrong of me to think that the right way to handle this is to change the instrumentation so it doesn't use a bad word?

Reply:
We can change the instrumentation sure, but also glean_parser should ideally to warn that we are using a forbidden word. Is it possible to know what are the forbidden words in C/C++ at the linter level? This gets hard easy... Ugh, why doesn't C++ have the nice little `r#` thingy like Rust does. So much easier /o\

Reply:
Eh, I filed a ticket. I am sure you all will come up with a good discussion. For now I'll keep my underscores. I really don't want to change any of the metrics we have right now https://bugzilla.mozilla.org/show_bug.cgi?id=1800904

Reply:
Well, the compiler knows what the forbidden words are, since it's what forbids their use. `errno` is fun because it's a macro that gets evaluated. `glean_parser` knowing the list of reserved words for the language it's outputting... I mean, we could. Each of the `.py` outputters could have a const list of all known bad words for the specific versions of the specific languages are we support.

Of course that'd mean glean_parser_ext would have to do it itself for languages like C++ that glean_parser doesn't support.